src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CcdCaseUpdater.java

```diff
@@ -0,0 +1,189 @@
+
+@Service
+public class CcdCaseUpdater {
+    private static final Logger log = LoggerFactory.getLogger(CcdCaseUpdater.class);
+
+    private final CaseUpdateClient caseUpdateClient;
+    private final AuthTokenGenerator s2sTokenGenerator;
+    private final CoreCaseDataApi coreCaseDataApi;
+    private final ExceptionRecordFinalizer exceptionRecordFinalizer;
+
+    public CcdCaseUpdater(
+        CaseUpdateClient caseUpdateClient,
+        AuthTokenGenerator s2sTokenGenerator,
+        CoreCaseDataApi coreCaseDataApi,
+        ExceptionRecordFinalizer exceptionRecordFinalizer
+    ) {
+        this.caseUpdateClient = caseUpdateClient;
+        this.s2sTokenGenerator = s2sTokenGenerator;
+        this.coreCaseDataApi = coreCaseDataApi;
+        this.exceptionRecordFinalizer = exceptionRecordFinalizer;
+    }
+
+    @SuppressWarnings({"squid:S2139", "unchecked"}) // squid for exception handle + logging
+    public ProcessResult updateCase(
+        ExceptionRecord exceptionRecord,
+        ServiceConfigItem configItem,
+        boolean ignoreWarnings,
+        String idamToken,
+        String userId,
+        CaseDetails existingCase
+    ) {
+        log.info(
+            "Start updating case for service {} with case ID {} from exception record {}",
+            configItem.getService(),
+            existingCase.getId(),
+            exceptionRecord.id
+        );
+
+        try {
+            String s2sToken = s2sTokenGenerator.generate();
+
+            SuccessfulUpdateResponse updateResponse = caseUpdateClient.updateCase(
+                configItem.getUpdateUrl(),
+                existingCase,
+                exceptionRecord,
+                s2sToken
+            );
+
+            if (!ignoreWarnings && !updateResponse.warnings.isEmpty()) {
+                // do not log warnings
+                return new ProcessResult(updateResponse.warnings, emptyList());
+            }
+
+            log.info(
+                "Successfully called case update endpoint of service {} to update case with case ID {} "
+                    + "based on exception record ref {}",
+                configItem.getService(),
+                existingCase.getId(),
+                exceptionRecord.id
+            );
+
+            updateCaseInCcd(
+                ignoreWarnings,
+                idamToken,
+                s2sToken,
+                userId,
+                exceptionRecord,
+                updateResponse.caseDetails,
+                existingCase
+            );
+
+            log.info(
+                "Successfully updated case for service {} with case ID {} based on exception record ref {}",
+                configItem.getService(),
+                existingCase.getId(),
+                exceptionRecord.id
+            );
+
+            return new ProcessResult(
+                exceptionRecordFinalizer.finalizeExceptionRecord(existingCase.getData(), existingCase.getId())
+            );
```

Comment:
It would be ideal if this code was wrapped in an `else`. It reads nicest when we have some sort of a flow tree, rather than a single line with optional returns. We could also extract the selected bit into a separate method, called `updateCaseInCCD`, for example.

Reply:
I have added 'else' clause. There is already 'updateCaseInCcd' method,, if we include log.info(...) calls into it we will need to provide service name as the value of a new parameter - but adding a new parameter just to log its value is not the best decision.

Reply:
Wouldn't hurt much, in my opinion. Logging is an important part of what the app does, and this would be a private method.

Reply:
ok